samples/nfc/tnep_poller/src/main.c

```diff
@@ -0,0 +1,696 @@
+
+		err = nfc_t4t_hl_procedure_cc_select();
+		if (err) {
+			printk("NFC T4T Capability Container select error: %d.\n",
+				err);
+		}
+
+		break;
+
+	case NFC_T4T_HL_PROCEDURE_CC_SELECT:
+		printk("NFC T4T Capability Container file selected.\n");
+
+		err = nfc_t4t_hl_procedure_cc_read(&NFC_T4T_CC_DESC(t4t_cc));
+		if (err) {
+			printk("Capability Container read error: %d.\n", err);
+		}
+
+		break;
+
+	case NFC_T4T_HL_PROCEDURE_NDEF_FILE_SELECT:
+		if (tnep_mode) {
+			printk("NFC T4T NDEF file contains the TNEP Initial Message selected.\n");
+
+			/* Select first service. */
+			err = nfc_tnep_poller_svc_select(tnep_rx_data,
+							 sizeof(tnep_rx_data),
+							 services, tnep_msg_max_size);
+			if (err) {
+				printk("Service select err: %d\n", err);
+			}
+
+			return;
+		}
+
+		printk("NFC T4T NDEF file selected.\n");
+
+		err = nfc_t4t_hl_procedure_ndef_read(&NFC_T4T_CC_DESC(t4t_cc),
+						     t4t.ndef[t4t.tlv_index],
+						     NFC_T4T_APDU_MAX_SIZE);
+		if (err) {
+			printk("NFC T4T NDEF file read error %d.\n", err);
+		}
+
+		break;
+
+	default:
+		break;
+	}
+
+	if (err) {
+		st25r3911b_nfca_tag_sleep();
+
+		k_delayed_work_submit(&transmit_work, TRANSMIT_DELAY);
+	}
+}
+
+static void t4t_hl_cc_read(struct nfc_t4t_cc_file *cc)
+{
+	int err;
+	struct nfc_t4t_tlv_block *tlv_block;
+
+	printk("NFC T4T Capability Container file read.\n");
+
+	for (size_t i = 0; i < cc->tlv_count; i++) {
+		tlv_block = &cc->tlv_block_array[i];
+
+		if ((tlv_block->type == NFC_T4T_TLV_BLOCK_TYPE_NDEF_FILE_CONTROL_TLV) &&
+		    (tlv_block->value.read_access == NFC_T4T_TLV_BLOCK_CONTROL_FILE_READ_ACCESS_GRANTED)) {
+			err = nfc_t4t_hl_procedure_ndef_file_select(tlv_block->value.file_id);
+			if (err) {
+				printk("NFC T4T NDEF file select error: %d.\n",
+				       err);
+			}
+
+			return;
+		}
+	}
+
+	printk("No NDEF File TLV in Capability Container.");
+}
+
+static void t4t_hl_ndef_read(u16_t file_id, const u8_t *data, size_t len)
+{
+	int err;
+	struct nfc_t4t_cc_file *cc;
+	struct nfc_t4t_tlv_block *tlv_block;
+
+	printk("NDEF file read, id: 0x%x.\n", file_id);
+
+	if (tnep_mode) {
+		err = nfc_tnep_poller_on_ndef_read(data + TAG_TYPE_4_NLEN_FIELD_SIZE,
+						   len - TAG_TYPE_4_NLEN_FIELD_SIZE);
+		if (err) {
+			printk("TNEP Read data error: %d\n", err);
+		}
+
+		return;
+	}
+
+	t4t.tlv_index++;
+	cc = &NFC_T4T_CC_DESC(t4t_cc);
+
+	for (size_t i = t4t.tlv_index; i < cc->tlv_count; i++) {
+		tlv_block = &cc->tlv_block_array[i];
+
+		if ((tlv_block->type == NFC_T4T_TLV_BLOCK_TYPE_NDEF_FILE_CONTROL_TLV) &&
+		    (tlv_block->value.read_access == NFC_T4T_TLV_BLOCK_CONTROL_FILE_READ_ACCESS_GRANTED)) {
+
+			err = nfc_t4t_hl_procedure_ndef_file_select(tlv_block->value.file_id);
+			if (err) {
+				printk("NFC T4T NDEF file select error: %d.\n",
+				       err);
+			}
+
+			return;
+		}
+
+		t4t.tlv_index++;
+	}
+
+	nfc_t4t_cc_file_printout(cc);
+
+	tlv_block = cc->tlv_block_array;
+	for (size_t i = 0; i < cc->tlv_count; i++) {
```

Comment:
First loop is used to read all NDEF message and print them after it finish its work the second loop looks for first message which contains TNEP Initial NDEF message. The first loop`s counter starts from last read message because we look for next message which can be read .

Reply:
Please  add some comment to explain the need of this procedure.